README.md
pyramid_oauth2_provider README
==============================

<aside class="warning">
Warning: You will need to reset your client secrets and provide a new ini 
config 'oauth2_provider.salt' when upgrading from v0.2.0.
</aside>

Getting Started
---------------

In an existing pyramid project you can take advantage of pyramid_oauth2_provider
by doing the following:

* Add `config.include('pyramid_oauth2_provider')` to your project setup. This
  will configure a `/oauth2/token` route for the token endpoint and an
  authentication policy that will support oauth2. If you want to be able to use 
  both cookie auth and oauth2 at the same time, you should use the
  `pyramid_oauth2_provider.authentication.OauthTktAuthenticationPolicy` instead
  of the default.
* Define a implementation of the `pyramid_oauth2_provider.interfaces.IAuthCheck`
  interface that works against your current user authentication check mechanism.
* In your paster configuration configure which IAuthCheck implementation to use
  by specifying `oauth2_provider.auth_checker`.
* In your production/development configuration, set a 16 random byte, base64 
  encoded salt for scrypt:
        
        oauth2_provider.salt = REPLACEME
        
  How to generate a salt in Python:
  
        from base64 import b64encode
        b64encode(os.urandom(16)).decode('utf-8')

* In your development configuration, you may also want to disable ssl
  enforcement by specifying `oauth2_provider.require_ssl = false`.
* Generate client credentials using the `create_client_credentials` script,
  provided as part of `pyramid_oauth2_provider`.

Request Flow
------------
Let's start by laying out a few ground rules when it comes to oauth2:

1. All requests *must* be made via HTTPS.
2. All data is transferred in headers and the body of messages rather than
   through url parameters.

The token endpoint is provided as a way to obtain and renew `access_tokens`.

#### Example initial token request:

        POST /oauth2/token HTTP/1.1
        Host: server.example.com
        Authorization: Basic czZCaGRSa3F0MzpnWDFmQmF0M2JW
        Content-Type: application/x-www-form-urlencoded

        grant_type=password&username=johndoe&password=A3ddj3w

* The basic auth header is the `client_id:client_secret` base64 encoded.
* Content-Type must be application/x-www-form-urlencoded

#### Example refresh token request:

        POST /token HTTP/1.1
        Host: server.example.com
        Authorization: Basic czZCaGRSa3F0MzpnWDFmQmF0M2JW
        Content-Type: application/x-www-form-urlencoded

        grant_type=refresh_token&refresh_token=tGzv3JOkF0XG5Qx2TlKW&user_id=1234

* The basic auth header is the `client_id:client_secret` base64 encoded.
* Content-Type must be application/x-www-form-urlencoded
* The `grant_type` must be "refresh".
* All form elements are required.

#### Example token response:

        HTTP/1.1 200 OK
        Content-Type: application/json;charset=UTF-8
        Cache-Control: no-store
        Pragma: no-cache

        {
          "access_token":"2YotnFZFEjr1zCsicMWpAA",
          "token_type":"bearer",
          "expires_in":3600,
          "refresh_token":"tGzv3JOkF0XG5Qx2TlKW",
          "user_id":1234,
        }

* The same response is returned for both auth token and refresh token requests.
* The `token_type` will always be "bearer".
* For purposes of this example the `access_token` and `refresh_token` are
  shorter than normal.


CHANGES.md
0.3.0 (TBD)
===========

- Python 3 support (#9)
- Removed epdb from create_client_credntials.py (#10)
- Fixed TypeError in OauthAuthenticationPolicy._get_auth_token (#14)
- Added scrypt hashing/salting of client secret - 
  new requirement: cryptography (#16)
- Changed/deprecated isRevoked() to is_revoked() in Oauth2Client, Oauth2Code, 
  and Oauth2Token (#16)
- Changed/deprecated asJSON() to as_json() in Oauth2Token (#16)
- Changed String DB columns to Unicode (#16)
- Added CHANGES.md and requirements.txt (#16)
- Code style fixes (#16)

pyramid_oauth2_provider/tests.py
#
# Copyright (c) Elliot Peele <elliot@bentlogic.net>
#
# This program is distributed under the terms of the MIT License as found
# in a file called LICENSE. If it is not present, the license
# is always available at http://www.opensource.org/licenses/mit-license.php.
#
# This program is distributed in the hope that it will be useful, but
# without any warrenty; without even the implied warranty of merchantability
# or fitness for a particular purpose. See the MIT License for full details.
#

import base64
import unittest
import transaction
from six.moves.urllib.parse import urlparse
from six.moves.urllib.parse import parse_qsl

from sqlalchemy import create_engine

from zope.interface import implementer

from pyramid import testing
from pyramid.response import Response

from . import jsonerrors
from .views import oauth2_token
from .views import oauth2_authorize
from .models import DBSession
from .models import Oauth2Token
from .models import Oauth2Client
from .models import Oauth2Code
from .models import Oauth2RedirectUri
from .models import initialize_sql
from .interfaces import IAuthCheck

_auth_value = None
_redirect_uri = None


@implementer(IAuthCheck)
class AuthCheck(object):
    def checkauth(self, username, password):
        return _auth_value


class TestCase(unittest.TestCase):
    def setUp(self):
        self.config = testing.setUp(
            settings={'oauth2_provider.salt': 'r+H5LT6EvgSSKFMZ2brdzQ=='}
        )
        self.config.registry.registerUtility(AuthCheck, IAuthCheck)

        engine = create_engine('sqlite://')
        initialize_sql(engine, self.config)

        self.auth = 1

        self.redirect_uri = 'http://localhost'

    def _get_auth(self):
        global _auth_value
        return _auth_value

    def _set_auth(self, value):
        global _auth_value
        _auth_value = value

    auth = property(_get_auth, _set_auth)

    def _get_redirect_uri(self):
        global _redirect_uri
        return _redirect_uri

    def _set_redirect_uri(self, uri):
        global _redirect_uri
        _redirect_uri = uri

    redirect_uri = property(_get_redirect_uri, _set_redirect_uri)

    def tearDown(self):
        DBSession.remove()
        testing.tearDown()

    def getAuthHeader(self, username, password, scheme='Basic'):
        encoded = base64.b64encode(('%s:%s' % (username, password)
                                    ).encode('utf-8'))
        return {'Authorization': '%s %s' % (scheme, encoded.decode('utf-8'))}


class TestAuthorizeEndpoint(TestCase):
    def setUp(self):
        TestCase.setUp(self)
        self.client = self._create_client()
        self.request = self._create_request()
        self.config.testing_securitypolicy(self.auth)

    def tearDown(self):
        TestCase.tearDown(self)
        self.client = None
        self.request = None

    def _create_client(self):
        with transaction.manager:
            client = Oauth2Client()
            DBSession.add(client)
            client_id = client.client_id

            redirect_uri = Oauth2RedirectUri(client, self.redirect_uri)
            DBSession.add(redirect_uri)

        client = DBSession.query(Oauth2Client).filter_by(client_id=client_id
                                                         ).first()
        return client

    def _create_request(self):
        data = {
            'response_type': 'code',
            'client_id': self.client.client_id
        }

        request = testing.DummyRequest(params=data)
        request.scheme = 'https'

        return request

    def _create_implicit_request(self):
        data = {
            'response_type': 'token',
            'client_id': self.client.client_id
        }

        request = testing.DummyRequest(post=data)
        request.scheme = 'https'

        return request

    def _process_view(self):
        with transaction.manager:
            token = oauth2_authorize(self.request)
        return token

    def _validate_authcode_response(self, response):
        self.assertTrue(isinstance(response, Response))
        self.assertEqual(response.status_int, 302)

        redirect = urlparse(self.redirect_uri)
        location = urlparse(response.location)
        self.assertEqual(location.scheme, redirect.scheme)
        self.assertEqual(location.hostname, redirect.hostname)
        self.assertEqual(location.path, redirect.path)
        self.assertFalse(location.fragment)

        params = dict(parse_qsl(location.query))

        self.assertTrue('code' in params)

        dbauthcodes = DBSession.query(Oauth2Code).filter_by(
            authcode=params.get('code')).all()

        self.assertTrue(len(dbauthcodes) == 1)

    def testAuthCodeRequest(self):
        response = self._process_view()
        self._validate_authcode_response(response)

    def testInvalidScheme(self):
        self.request.scheme = 'http'
        response = self._process_view()
        self.assertTrue(isinstance(response, jsonerrors.HTTPBadRequest))

    def testDisableSchemeCheck(self):
        self.request.scheme = 'http'
        self.config.get_settings()['oauth2_provider.require_ssl'] = False
        response = self._process_view()
        self._validate_authcode_response(response)

    def testNoClientCreds(self):
        self.request.params.pop('client_id')
        response = self._process_view()
        self.assertTrue(isinstance(response, jsonerrors.HTTPBadRequest))

    def testNoResponseType(self):
        self.request.params.pop('response_type')
        response = self._process_view()
        self.assertTrue(isinstance(response, jsonerrors.HTTPBadRequest))

    def testRedirectUriSupplied(self):
        self.request.params['redirect_uri'] = self.redirect_uri
        response = self._process_view()
        self._validate_authcode_response(response)

    def testMultipleRedirectUrisUnspecified(self):
        with transaction.manager:
            redirect_uri = Oauth2RedirectUri(
                self.client, 'https://otherhost.com')
            DBSession.add(redirect_uri)
        response = self._process_view()
        self.assertTrue(isinstance(response, jsonerrors.HTTPBadRequest))

    def testMultipleRedirectUrisSpecified(self):
        with transaction.manager:
            redirect_uri = Oauth2RedirectUri(
                self.client, 'https://otherhost.com')
            DBSession.add(redirect_uri)
        self.request.params['redirect_uri'] = 'https://otherhost.com'
        self.redirect_uri = 'https://otherhost.com'
        response = self._process_view()
        self._validate_authcode_response(response)

    def testRetainRedirectQueryComponent(self):
        uri = 'https://otherhost.com/and/path?some=value'
        with transaction.manager:
            redirect_uri = Oauth2RedirectUri(
                self.client, uri)
            DBSession.add(redirect_uri)
        self.request.params['redirect_uri'] = uri
        self.redirect_uri = uri
        response = self._process_view()
        self._validate_authcode_response(response)

        parts = urlparse(response.location)
        params = dict(parse_qsl(parts.query))

        self.assertTrue('some' in params)
        self.assertEqual(params['some'], 'value')

    def testState(self):
        state_value = 'testing'
        self.request.params['state'] = state_value
        response = self._process_view()
        self._validate_authcode_response(response)
        parts = urlparse(response.location)
        params = dict(parse_qsl(parts.query))
        self.assertTrue('state' in params)
        self.assertEqual(state_value, params['state'])


class TestTokenEndpoint(TestCase):
    def setUp(self):
        TestCase.setUp(self)
        self.client, self.client_secret = self._create_client()
        self.request = self._create_request()

    def tearDown(self):
        TestCase.tearDown(self)
        self.client = None
        self.request = None

    def _create_client(self):
        with transaction.manager:
            client = Oauth2Client()
            client_secret = client.new_client_secret()
            DBSession.add(client)
            client_id = client.client_id

        client = DBSession.query(Oauth2Client).filter_by(
            client_id=client_id).first()
        return client, client_secret

    def _create_request(self):
        headers = self.getAuthHeader(
            self.client.client_id,
            self.client_secret)

        data = {
            'grant_type': 'password',
            'username': 'john',
            'password': 'foo',
        }

        request = testing.DummyRequest(post=data, headers=headers)
        request.scheme = 'https'

        return request

    def _create_refresh_token_request(self, refresh_token, user_id):
        headers = self.getAuthHeader(
            self.client.client_id,
            self.client_secret)

        data = {
            'grant_type': 'refresh_token',
            'refresh_token': refresh_token,
            'user_id': str(user_id),
        }

        request = testing.DummyRequest(post=data, headers=headers)
        request.scheme = 'https'

        return request

    def _process_view(self):
        with transaction.manager:
            token = oauth2_token(self.request)
        return token

    def _validate_token(self, token):
        self.assertTrue(isinstance(token, dict))
        self.assertEqual(token.get('user_id'), self.auth)
        self.assertEqual(token.get('expires_in'), 3600)
        self.assertEqual(token.get('token_type'), 'bearer')
        self.assertEqual(len(token.get('access_token')), 64)
        self.assertEqual(len(token.get('refresh_token')), 64)
        self.assertEqual(len(token), 5)

        dbtoken = DBSession.query(Oauth2Token).filter_by(
            access_token=token.get('access_token')).first()

        self.assertEqual(dbtoken.user_id, token.get('user_id'))
        self.assertEqual(dbtoken.expires_in, token.get('expires_in'))
        self.assertEqual(dbtoken.access_token, token.get('access_token'))
        self.assertEqual(dbtoken.refresh_token, token.get('refresh_token'))

    def testTokenRequest(self):
        self.auth = 500
        token = self._process_view()
        self._validate_token(token)

    def testInvalidMethod(self):
        self.request.method = 'GET'
        token = self._process_view()
        self.assertTrue(isinstance(token, jsonerrors.HTTPMethodNotAllowed))

    def testInvalidScheme(self):
        self.request.scheme = 'http'
        token = self._process_view()
        self.assertTrue(isinstance(token, jsonerrors.HTTPBadRequest))

    def testDisableSchemeCheck(self):
        self.request.scheme = 'http'
        self.config.get_settings()['oauth2_provider.require_ssl'] = False
        token = self._process_view()
        self._validate_token(token)

    def testNoClientCreds(self):
        self.request.headers = {}
        token = self._process_view()
        self.assertTrue(isinstance(token, jsonerrors.HTTPUnauthorized))

    def testInvalidClientCreds(self):
        self.request.headers = self.getAuthHeader(
            self.client.client_id, 'abcde')
        token = self._process_view()
        self.assertTrue(isinstance(token, jsonerrors.HTTPBadRequest))

    def testInvalidGrantType(self):
        self.request.POST['grant_type'] = 'foo'
        token = self._process_view()
        self.assertTrue(isinstance(token, jsonerrors.HTTPBadRequest))

    def testCacheHeaders(self):
        self._process_view()
        self.assertEqual(
            self.request.response.headers.get('Cache-Control'), 'no-store')
        self.assertEqual(
            self.request.response.headers.get('Pragma'), 'no-cache')

    def testMissingUsername(self):
        self.request.POST.pop('username')
        token = self._process_view()
        self.assertTrue(isinstance(token, jsonerrors.HTTPBadRequest))

    def testMissingPassword(self):
        self.request.POST.pop('password')
        token = self._process_view()
        self.assertTrue(isinstance(token, jsonerrors.HTTPBadRequest))

    def testFailedPassword(self):
        self.auth = False
        token = self._process_view()
        self.assertTrue(isinstance(token, jsonerrors.HTTPUnauthorized))

    def testRefreshToken(self):
        token = self._process_view()
        self._validate_token(token)
        self.request = self._create_refresh_token_request(
            token.get('refresh_token'), token.get('user_id'))
        token = self._process_view()
        self._validate_token(token)

    def testMissingRefreshToken(self):
        token = self._process_view()
        self._validate_token(token)
        self.request = self._create_refresh_token_request(
            token.get('refresh_token'), token.get('user_id'))
        self.request.POST.pop('refresh_token')
        token = self._process_view()
        self.assertTrue(isinstance(token, jsonerrors.HTTPBadRequest))

    def testMissingUserId(self):
        token = self._process_view()
        self._validate_token(token)
        self.request = self._create_refresh_token_request(
            token.get('refresh_token'), token.get('user_id'))
        self.request.POST.pop('user_id')
        token = self._process_view()
        self.assertTrue(isinstance(token, jsonerrors.HTTPBadRequest))

    def testInvalidRefreshToken(self):
        token = self._process_view()
        self._validate_token(token)
        self.request = self._create_refresh_token_request(
            'abcd', token.get('user_id'))
        token = self._process_view()
        self.assertTrue(isinstance(token, jsonerrors.HTTPUnauthorized))

    def testRefreshInvalidClientId(self):
        token = self._process_view()
        self._validate_token(token)
        self.request = self._create_refresh_token_request(
            token.get('refresh_token'), token.get('user_id'))
        self.request.headers = self.getAuthHeader(
            '1234', self.client_secret)
        token = self._process_view()
        self.assertTrue(isinstance(token, jsonerrors.HTTPBadRequest))

    def testUserIdMissmatch(self):
        token = self._process_view()
        self._validate_token(token)
        self.request = self._create_refresh_token_request(
            token.get('refresh_token'), '2')
        token = self._process_view()
        self.assertTrue(isinstance(token, jsonerrors.HTTPBadRequest))

    def testRevokedAccessTokenRefresh(self):
        token = self._process_view()
        self._validate_token(token)

        dbtoken = DBSession.query(Oauth2Token).filter_by(
            access_token=token.get('access_token')).first()
        dbtoken.revoke()

        self.request = self._create_refresh_token_request(
            token.get('refresh_token'), token.get('user_id'))
        token = self._process_view()
        self._validate_token(token)

    def testTimeRevokeAccessToken(self):
        token = self._process_view()
        self._validate_token(token)

        dbtoken = DBSession.query(Oauth2Token).filter_by(
            access_token=token.get('access_token')).first()
        dbtoken.expires_in = 0

        self.assertEqual(dbtoken.isRevoked(), True)

    def testTimeRevokeAccessToken2(self):
        token = self._process_view()
        self._validate_token(token)

        dbtoken = DBSession.query(Oauth2Token).filter_by(
            access_token=token.get('access_token')).first()
        dbtoken.expires_in = 10

        self.assertEqual(dbtoken.isRevoked(), False)


pyramid_oauth2_provider/jsonerrors.py
#
# Copyright (c) Elliot Peele <elliot@bentlogic.net>
#
# This program is distributed under the terms of the MIT License as found
# in a file called LICENSE. If it is not present, the license
# is always available at http://www.opensource.org/licenses/mit-license.php.
#
# This program is distributed in the hope that it will be useful, but
# without any warrenty; without even the implied warranty of merchantability
# or fitness for a particular purpose. See the MIT License for full details.
#

"""
Custom HTTP exceptions that support rendering to JSON by default.
"""

# NOTE: If you need to add more errors, please subclass errors from
#       httpexceptions as has been done below.

from string import Template

from pyramid import httpexceptions
from pyramid.httpexceptions import text_type
from pyramid.httpexceptions import _no_escape
from pyramid.httpexceptions import WSGIHTTPException

def _quote_escape(value):
    v = _no_escape(value)
    return v.replace('"', '\\"')


class BaseJsonHTTPError(WSGIHTTPException):
    """
    Base error class for rendering errors in JSON.
    """

    json_template_obj = Template('''\
{
    "status": "${status}",
    "code": ${code},
    "explanation": "${explanation}",
    "detail": "${detail}"
}
${html_comment}
''')

    def prepare(self, environ):
        """
        Always return errors in JSON.
        """

        if not self.body and not self.empty_body:
            html_comment = ''
            comment = self.comment or ''
            accept = environ.get('HTTP_ACCEPT', '')
            if 'text/plain' in accept:
                self.content_type = 'text/plain'
                escape = _no_escape
                page_template = self.plain_template_obj
                br = '\n'
                if comment:
                    html_comment = escape(comment)
            else:
                self.content_type = 'aplication/json'
                escape = _quote_escape
                page_template = self.json_template_obj
                br = '\n'
                if comment:
                    html_comment = '# %s' % comment
            args = {
                'br': br,
                'explanation': escape(self.explanation),
                'detail': escape(self.detail or ''),
                'comment': escape(comment),
                'html_comment': html_comment,
                }
            for k, v in list(environ.items()):
                if (not k.startswith('wsgi.')) and ('.' in k):
                    continue
                args[k] = escape(v)
            for k, v in list(self.headers.items()):
                args[k.lower()] = escape(v)
            page = page_template.substitute(status=self.status,
                code=self.code, **args)
            if isinstance(page, text_type):
                if not self.charset:
                    self.charset = 'utf-8'
                page = page.encode(self.charset)
            self.app_iter = [page]
            self.body = page


class HTTPBadRequest(httpexceptions.HTTPBadRequest, BaseJsonHTTPError):
    pass


class HTTPUnauthorized(httpexceptions.HTTPUnauthorized, BaseJsonHTTPError):
    pass


class HTTPMethodNotAllowed(httpexceptions.HTTPMethodNotAllowed,
    BaseJsonHTTPError):
    pass


pyramid_oauth2_provider/views.py
#
# Copyright (c) Elliot Peele <elliot@bentlogic.net>
#
# This program is distributed under the terms of the MIT License as found
# in a file called LICENSE. If it is not present, the license
# is always available at http://www.opensource.org/licenses/mit-license.php.
#
# This program is distributed in the hope that it will be useful, but
# without any warranty; without even the implied warranty of merchantability
# or fitness for a particular purpose. See the MIT License for full details.
#

import logging

from base64 import b64decode

from pyramid.view import view_config
from pyramid.security import NO_PERMISSION_REQUIRED
from pyramid.security import authenticated_userid
from pyramid.security import Authenticated
from pyramid.httpexceptions import HTTPFound
from six.moves.urllib.parse import urlparse
from six.moves.urllib.parse import parse_qsl
from six.moves.urllib.parse import ParseResult
from six.moves.urllib.parse import urlencode
from cryptography.hazmat.primitives.kdf.scrypt import Scrypt
from cryptography.exceptions import InvalidKey

from .models import DBSession as db
from .models import Oauth2Token
from .models import Oauth2Code
from .models import Oauth2RedirectUri
from .models import Oauth2Client
from .models import backend
from .errors import InvalidToken
from .errors import InvalidClient
from .errors import InvalidRequest
from .errors import UnsupportedGrantType
from .util import oauth2_settings
from .util import getClientCredentials
from .interfaces import IAuthCheck
from .jsonerrors import HTTPBadRequest
from .jsonerrors import HTTPUnauthorized
from .jsonerrors import HTTPMethodNotAllowed


def require_https(handler):
    """
     This check should be taken care of via the authorization policy, but in
     case someone has configured a different policy, check again. HTTPS is
     required for all Oauth2 authenticated requests to ensure the security of
     client credentials and authorization tokens.
    """
    def wrapped(request):
        if (request.scheme != 'https' and
                oauth2_settings('require_ssl', default=True)):
            log.info('rejected request due to unsupported scheme: %s'
                     % request.scheme)
            return HTTPBadRequest(InvalidRequest(
                error_description='Oauth2 requires all requests to be made '
                                  'via HTTPS.'))
        return handler(request)
    return wrapped


log = logging.getLogger('pyramid_oauth2_provider.views')


@view_config(route_name='oauth2_provider_authorize', renderer='json',
             permission=Authenticated)
@require_https
def oauth2_authorize(request):
    """
    * In the case of a 'code' authorize request a GET or POST is made
    with the following structure.

        GET /authorize?response_type=code&client_id=aoiuer HTTP/1.1
        Host: server.example.com

        POST /authorize HTTP/1.1
        Host: server.example.com
        Content-Type: application/x-www-form-urlencoded

        response_type=code&client_id=aoiuer

    The response_type and client_id are required parameters. A redirect_uri
    and state parameters may also be supplied. The redirect_uri will be
    validated against the URI's registered for the client. The state is an
    opaque value that is simply passed through for security on the client's
    end.

    The response to a 'code' request will be a redirect to a registered URI
    with the authorization code and optional state values as query
    parameters.

        HTTP/1.1 302 Found
        Location: https://client.example.com/cb?code=AverTaer&state=efg

    """
    if not request.params.get('client_id'):
        return HTTPBadRequest(InvalidRequest(
            error_description='Invalid client credentials'))
    request.client_id = request.params.get('client_id')

    client = db.query(Oauth2Client).filter_by(
        client_id=request.client_id).first()

    if not client:
        log.info('received invalid client credentials')
        return HTTPBadRequest(InvalidRequest(
            error_description='Invalid client credentials'))

    redirect_uri = (request.params.get('redirect_uri') if
                    request.params.get('redirect_uri') else None)
    redirection_uri = None
    if len(client.redirect_uris) == 1 and (
                not redirect_uri or redirect_uri == client.redirect_uris[0]):
        redirection_uri = client.redirect_uris[0]
    elif len(client.redirect_uris) > 0:
        redirection_uri = db.query(Oauth2RedirectUri)\
            .filter_by(client_id=client.id, uri=redirect_uri).first()

    if redirection_uri is None:
        return HTTPBadRequest(InvalidRequest(
            error_description='Redirection URI validation failed'))

    response_type = request.params.get('response_type')
    state = request.params.get('state')
    if 'code' == response_type:
        resp = handle_authcode(request, client, redirection_uri, state)
    elif 'token' == response_type:
        resp = handle_implicit(request, client, redirection_uri, state)
    else:
        log.info('received invalid response_type %s')
        resp = HTTPBadRequest(InvalidRequest(
            error_description='Oauth2 unknown response_type not supported'))

    return resp


def handle_authcode(request, client, redirection_uri, state=None):
    parts = urlparse(redirection_uri.uri)
    qparams = dict(parse_qsl(parts.query))

    user_id = authenticated_userid(request)
    auth_code = Oauth2Code(client, user_id)
    db.add(auth_code)
    db.flush()

    qparams['code'] = auth_code.authcode
    if state:
        qparams['state'] = state
    parts = ParseResult(
        parts.scheme, parts.netloc, parts.path, parts.params,
        urlencode(qparams), '')
    return HTTPFound(location=parts.geturl())


def handle_implicit(request, client, redirection_uri, state=None):
    return HTTPBadRequest(InvalidRequest(
        error_description='Oauth2 response_type "implicit" not supported'))


@view_config(route_name='oauth2_provider_token', renderer='json',
             permission=NO_PERMISSION_REQUIRED)
@require_https
def oauth2_token(request):
    """
    * In the case of an incoming authentication request a POST is made
    with the following structure.

        POST /token HTTP/1.1
        Host: server.example.com
        Authorization: Basic czZCaGRSa3F0MzpnWDFmQmF0M2JW
        Content-Type: application/x-www-form-urlencoded

        grant_type=password&username=johndoe&password=A3ddj3w

    The basic auth header contains the client_id:client_secret base64
    encoded for client authentication.

    The username and password are form encoded as part of the body. This
    request *must* be made over https.

    The response to this request will be, assuming no error:

        HTTP/1.1 200 OK
        Content-Type: application/json;charset=UTF-8
        Cache-Control: no-store
        Pragma: no-cache

        {
          "access_token":"2YotnFZFEjr1zCsicMWpAA",
          "token_type":"bearer",
          "expires_in":3600,
          "refresh_token":"tGzv3JOkF0XG5Qx2TlKW",
          "user_id":1234,
        }

    * In the case of a token refresh request a POST with the following
    structure is required:

        POST /token HTTP/1.1
        Host: server.example.com
        Authorization: Basic czZCaGRSa3F0MzpnWDFmQmF0M2JW
        Content-Type: application/x-www-form-urlencoded

        grant_type=refresh_token&refresh_token=tGzv3JOkF0XG5Qx2TlKW&user_id=1234

    The response will be the same as above with a new access_token and
    refresh_token.
    """

    # Make sure this is a POST.
    if request.method != 'POST':
        log.info('rejected request due to invalid method: %s' % request.method)
        return HTTPMethodNotAllowed(
            'This endpoint only supports the POST method.')

    getClientCredentials(request)

    # Make sure we got a client_id and secret through the authorization
    # policy. Note that you should only get here if not using the Oauth2
    # authorization policy or access was granted through the AuthTKt policy.
    if (not hasattr(request, 'client_id') or
        not hasattr(request, 'client_secret')):
        log.info('did not receive client credentials')
        return HTTPUnauthorized('Invalid client credentials')

    client = db.query(Oauth2Client).filter_by(
        client_id=request.client_id).first()

    # Again, the authorization policy should catch this, but check again.
    if not oauth2_settings('salt'):
        raise ValueError('oauth2_provider.salt configuration required.')
    salt = b64decode(oauth2_settings('salt').encode('utf-8'))
    kdf = Scrypt(
        salt=salt,
        length=64,
        n=2 ** 14,
        r=8,
        p=1,
        backend=backend
    )

    try:
        client_secret = request.client_secret
        try:
            client_secret = bytes(client_secret, 'utf-8')
        except TypeError:
            client_secret = client_secret.encode('utf-8')
        kdf.verify(client_secret, client.client_secret)
        bad_secret = False
    except (AttributeError, InvalidKey):
        bad_secret = True
    if not client or bad_secret:
        log.info('received invalid client credentials')
        return HTTPBadRequest(InvalidRequest(
            error_description='Invalid client credentials'))

    # Check for supported grant type. This is a required field of the form
    # submission.
    grant_type = request.POST.get('grant_type')
    if grant_type == 'password':
        resp = handle_password(request, client)
    elif grant_type == 'refresh_token':
        resp = handle_refresh_token(request, client)
    else:
        log.info('invalid grant type: %s' % grant_type)
        return HTTPBadRequest(UnsupportedGrantType(
            error_description='Only password and refresh_token grant types are'
                              ' supported by this authentication server'))

    add_cache_headers(request)
    return resp


def handle_password(request, client):
    if 'username' not in request.POST or 'password' not in request.POST:
        log.info('missing username or password')
        return HTTPBadRequest(InvalidRequest(
            error_description='Both username and password are required to '
                              'obtain a password based grant.'))

    auth_check = request.registry.queryUtility(IAuthCheck)
    user_id = auth_check().checkauth(
        request.POST.get('username'),
        request.POST.get('password'))

    if not user_id:
        log.info('could not validate user credentials')
        return HTTPUnauthorized(InvalidClient(
            error_description='Username and password are invalid.'))

    auth_token = Oauth2Token(client, user_id)
    db.add(auth_token)
    db.flush()
    return auth_token.asJSON(token_type='bearer')


def handle_refresh_token(request, client):
    if 'refresh_token' not in request.POST:
        log.info('refresh_token field missing')
        return HTTPBadRequest(InvalidRequest(
            error_description='refresh_token field required'))

    if 'user_id' not in request.POST:
        log.info('user_id field missing')
        return HTTPBadRequest(InvalidRequest(
            error_description='user_id field required'))

    auth_token = db.query(Oauth2Token).filter_by(
        refresh_token=request.POST.get('refresh_token')
    ).first()

    if not auth_token:
        log.info('invalid refresh_token')
        return HTTPUnauthorized(InvalidToken(
            error_description='Provided refresh_token is not valid.'))

    if auth_token.client.client_id != client.client_id:
        log.info('invalid client_id')
        return HTTPBadRequest(InvalidClient(
            error_description='Client does not own this refresh_token.'))

    if str(auth_token.user_id) != request.POST.get('user_id'):
        log.info('invalid user_id')
        return HTTPBadRequest(InvalidClient(
            error_description='The given user_id does not match the given '
                              'refresh_token.'))

    new_token = auth_token.refresh()
    db.add(new_token)
    db.flush()
    return new_token.asJSON(token_type='bearer')


def add_cache_headers(request):
    """
    The Oauth2 draft spec requires that all token endpoint traffic be marked
    as uncacheable.
    """

    resp = request.response
    resp.headerlist.append(('Cache-Control', 'no-store'))
    resp.headerlist.append(('Pragma', 'no-cache'))
    return request
